pints/_optimisers/_hager_zhang_line_search.py

```diff
@@ -0,0 +1,899 @@
+
+        # As c1 approaches 0 and c2 approaches 1, the line search
+        # terminates more quickly.
+        self._c1 = 1E-4  # Parameter for Armijo condition rule, 0 < c1 < 0.5
+        self._c2 = 0.9  # Parameter for curvature condition rule, c1 < c2 < 1.0
+
+        # boundary values of alpha
+        self._minimum_alpha = 0.0
+        self._maximum_alpha = float("inf")
+        self._proposed_alpha = 0.001  # same default value as used in stan
+
+        # Increase allowed between accepted positions when using approximate
+        # wolfe conditions, this takes into acccount machine error and
+        # insures decreasing.
+        self.epsilon = 1E-6
+
+        # range (0, 1), used in the ``self.__update()`` and
+        # ``self.__initial_bracket()`` when the potential intervals violate
+        # the opposite slope condition (see function definition)
+        self.theta = 0.5
+
+        self.__gamma = 0.66
+
+        # range (0, 1) small factor used in initial guess of step size
+        self.__ps_0 = 0.01
+        # range (0, 1) small factor used in subsequent guesses of step size
+        self.__ps_1 = 0.1
+        # range (1, inf) factor used in subsequent guesses of step size
+        self.__ps_2 = 2.0
+
+        self.__current_alpha = None
+
+        # approximate inverse hessian
+        # initial the identity is used
+        self._B = np.identity(self._n_parameters)
+
+        # newton direction
+        self.__px = None
+
+        # number of accepted steps/ newton direction updates
+        self._k = 0
+
+        # number of steps in the current line search iteration
+        self.__j = 0
+        self.__logic_steps_left = 0
+
+        # maximum number of line search steps before a successful point
+
+        # packed reply of objective function and gradient evaluations
+        # needed by generator implementation
+        # this approach is taken rather than the send() keyword to
+        # back date past Python 2.5
+        self.__reply_f_and_dfdx = None
+
+        # Current point, score, and gradient
+        self._current = self._x0
+        self._current_f = None
+        self._current_dfdx = None
+
+        # Proposed next point (read-only, so can be passed to user)
+        self._proposed = self._x0
+        self._proposed.setflags(write=False)
+
+        self.__convergence = False
+
+        # ask generator
+        self.__ask_generator = self.__ask_generator()
+
+        # need to check wolfe conditions
+        self.wolfe_check = False
+        # continue main body of line search loop
+        self.__line_search_looping = None
+        # wheather the wolfe conditions have been meet
+        self.__wolfe_cond_meet = False
+
+    def name(self):
+        """ See :meth:`Optimiser.name()`. """
+        return 'Hager-Zhang Line Search'
+
+    def fbest(self):
+        """ See :meth:`Optimiser.fbest()`. """
+        return self._fbest
+
+    def wolfe_line_search_parameters(self):
```

Comment:
Ah ha -- you've already bundled them here. How about calling them something like: `wolfe_line_param_1` and `wolfe_line_param_2`?